File: Spark_Datastream/task1.py
import json
import sys
from itertools import combinations
import math
import random
import binascii
from blackbox import BlackBox


BIT_ARRAY_SIZE = 69997
NUM_HASH_FUNCTIONS = 10
bit_array = [0] * BIT_ARRAY_SIZE
true_users = set()


# ((ax + b) % p) % m
def generate_hash(m = BIT_ARRAY_SIZE, num_hush = NUM_HASH_FUNCTIONS):

    prime_list = [50021, 50023, 50033, 50047, 50051, 50069, 50077, 50087, 50093, 50101]
    # prime_list = [5003, 5009, 5011, 5021, 5023, 5039, 5051, 5059, 5077, 5081]
    hash_functions=[]

    for _ in range(num_hush):
        p = random.choice(prime_list)
        a = random.randint(1, p-1)
        b = random.randint(1, p-1)

        def hash_func(x, a=a, b=b, p=p):
            return ((a * x + b) % p) % m
        
        hash_functions.append(hash_func)
    
    return hash_functions

global_hash_functions = generate_hash()


# get the result of hash 
def myhashs(s):
    hash_results = []
    input_number = int(binascii.hexlify(s.encode('utf8')), 16)

    for hash_func in global_hash_functions:
        hash_results.append(hash_func(input_number))

    return hash_results
    


def main(input_file, stream_size, num_of_asks, output_file):

    bx = BlackBox()
    hash_functions = generate_hash()

    with open(output_file, 'w') as f_out:
        f_out.write("Time,FPR\n")

        for i in range(num_of_asks):
            user_steam = bx.ask(input_file, stream_size)

            false_positive = 0
            total_negative = 0

            for user_id in user_steam:
                hashes = myhashs(user_id)
                seen = all(bit_array[h] == 1 for h in hashes)

                if seen and user_id not in true_users:
                    false_positive += 1

                if user_id not in true_users:
                    total_negative += 1
                    true_users.add(user_id)
                    for h in hashes:
                        bit_array[h] = 1

            fpr = false_positive / total_negative if total_negative > 0 else 0

            f_out.write(f"{i},{fpr}\n")



if __name__ == "__main__":
    input_file = sys.argv[1]
    stream_size = int(sys.argv[2])
    num_of_asks = int(sys.argv[3])
    output_file = sys.argv[4]
    main(input_file, stream_size, num_of_asks, output_file)

File: Spark_Datastream/task1.scala
import scala.io.Source
import scala.util.Random
import scala.collection.mutable
import java.io.PrintWriter
import java.io.File

import Blackbox._  

object task1 {


  val BIT_ARRAY_SIZE: Int = 69997         
  val NUM_HASH_FUNCTIONS: Int = 10      


  val bitArray: Array[Int] = Array.fill(BIT_ARRAY_SIZE)(0)

  val trueUsers: mutable.Set[String] = mutable.Set.empty[String]


  def generateHash(m: Int = BIT_ARRAY_SIZE, numHush: Int = NUM_HASH_FUNCTIONS): List[BigInt => Int] = {
    val primeList = List(50021, 50023, 50033, 50047, 50051, 50069, 50077, 50087, 50093, 50101)
    val hashFunctions = (1 to numHush).map { _ =>
      val p = primeList(Random.nextInt(primeList.length))

      val a = Random.nextInt(p - 1) + 1
      val b = Random.nextInt(p - 1) + 1

      (x: BigInt) => (((a * x + b) % p) % m).toInt
    }.toList
    hashFunctions
  }

  // Global hash functions generated once
  val globalHashFunctions: List[BigInt => Int] = generateHash()


  def myhashs(s: String): List[Int] = {
    val inputNumber = BigInt(s.getBytes("UTF-8"))
    globalHashFunctions.map(hf => hf(inputNumber))
  }

  def main(args: Array[String]): Unit = {
    if (args.length != 4) {
      sys.exit(1)
    }

    val inputFile = args(0)
    val streamSize = args(1).toInt
    val numOfAsks = args(2).toInt
    val outputFile = args(3)


    val bx = Blackbox()
    
    val pw = new PrintWriter(new File(outputFile))
    pw.println("Time,FPR")

    // Process each round of the data stream
    for (i <- 0 until numOfAsks) {
     
      val userStream: Array[String] = bx.ask(inputFile, streamSize)
      var falsePositive: Double = 0.0  
      var totalNegative: Double = 0.0 

      for (userId <- userStream) {
        // Use the global hash functions to compute hash values for the user
        val hashes: List[Int] = myhashs(userId)
        // Check if all positions in the bit array corresponding to these hash values are set to 1
        val seen: Boolean = hashes.forall(h => bitArray(h) == 1)
        if (seen && !trueUsers.contains(userId)) {
          falsePositive += 1.0
        }

        // If the user has not been seen before, update the counter, trueUsers set, and the bit array.
        if (!trueUsers.contains(userId)) {
          totalNegative += 1.0
          trueUsers.add(userId)
          hashes.foreach { h =>
            bitArray(h) = 1
          }
        }
      }

      val fpr: Double = if (totalNegative > 0) falsePositive / totalNegative else 0.0
      pw.println(s"$i,$fpr")
    }

    pw.close()
  }
}
